File: e2e_tests/websockets/typescript/test_websockets.mjs
// End-to-end WebSocket test for the TypeScript / NAPI SDK.
//
// Mirrors `e2e_tests/websockets/rust/test_websockets.rs` and the Python
// suite — exercises the unified `WebSocket` class against both exchanges
// to verify the Node binding hasn't drifted.
//
// Run from repo root:
//     just e2e-typescript websockets
//
// Output: writes ../results/typescript.json + pass/fail summary, exits
// non-zero if any case fails.

import { Exchange, WebSocket } from '../../../sdks/typescript/index.js';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_ROOT = path.resolve(__dirname, '../../..');
const RESULTS_DIR = path.join(REPO_ROOT, 'e2e_tests', 'websockets', 'results');
fs.mkdirSync(RESULTS_DIR, { recursive: true });
const RESULTS_PATH = path.join(RESULTS_DIR, 'typescript.json');

function loadDotenv(p) {
    if (!fs.existsSync(p)) return;
    for (const line of fs.readFileSync(p, 'utf8').split('\n')) {
        const t = line.trim();
        if (!t || t.startsWith('#') || !t.includes('=')) continue;
        const [k, ...rest] = t.split('=');
        const v = rest.join('=');
        if (process.env[k.trim()] === undefined) process.env[k.trim()] = v.trim();
    }
}
loadDotenv(path.join(REPO_ROOT, '.env'));

const keyPath = process.env.KALSHI_PRIVATE_KEY_PATH;
if (keyPath && !path.isAbsolute(keyPath)) {
    process.env.KALSHI_PRIVATE_KEY_PATH = path.join(REPO_ROOT, keyPath);
}

function trim(o) {
    return Object.fromEntries(Object.entries(o).filter(([, v]) => v));
}

function configFor(eid) {
    return eid === 'kalshi'
        ? trim({
            api_key_id: process.env.KALSHI_API_KEY_ID,
            private_key_path: process.env.KALSHI_PRIVATE_KEY_PATH,
        })
        : trim({
            private_key: process.env.POLYMARKET_PRIVATE_KEY,
            funder: process.env.POLYMARKET_FUNDER,
            signature_type: process.env.POLYMARKET_SIGNATURE_TYPE,
            api_key: process.env.POLYMARKET_API_KEY,
            api_secret: process.env.POLYMARKET_API_SECRET,
            api_passphrase: process.env.POLYMARKET_API_PASSPHRASE,
        });
}

const makeExchange = (eid) => new Exchange(eid, configFor(eid));
const makeWs = (eid) => new WebSocket(eid, configFor(eid));

// -----------------------------------------------------------------------
// Result tracking
// -----------------------------------------------------------------------

const results = [];

function record(name, expect, status, detail) {
    const entry = { name, expect, status, detail };
    results.push(entry);
    const sym = { pass: '✓', fail: '✗', skip: '·' }[status] || '?';
    console.log(`  ${sym} [${status.toUpperCase()}] ${name}`);
    if (status === 'fail' && detail) console.log(`      → ${String(detail).slice(0, 200)}`);
}

async function caseAsync(name, fn, expect = 'ok') {
    try {
        const r = await fn();
        if (r === 'skip') return record(name, expect, 'skip', 'skipped by test');
        record(name, expect, expect === 'err' ? 'fail' : 'pass', null);
    } catch (e) {
        const msg = String(e?.message || e).toLowerCase();
        if (expect === 'err') return record(name, expect, 'pass', `expected err: ${e}`);
        if (msg.includes('rate limit') || msg.includes('429') || msg.includes('timed out') || msg.includes('http error')) {
            return record(name, expect, 'skip', `transient: ${e}`);
        }
        record(name, expect, 'fail', `raised: ${e}`);
    }
}

// -----------------------------------------------------------------------
// Helpers
// -----------------------------------------------------------------------

async function discoverActiveMarket(eid) {
    const ex = makeExchange(eid);
    const page = await ex.fetchMarkets('active', undefined, undefined, undefined, undefined, 50);
    const markets = Array.isArray(page) ? page : page.markets || page;
    for (const m of markets.slice(0, 20)) {
        const token = (m.outcomes || []).map((o) => o.token_id).find(Boolean);
        const ticker = m.ticker;
        const assetId = token || ticker;
        if (!assetId) continue;
        try {
            const book = await ex.fetchOrderbook(assetId);
            if ((book.bids || []).length > 0 || (book.asks || []).length > 0) {
                return { subId: eid === 'kalshi' ? ticker : assetId, assetId };
            }
        } catch { /* keep trying */ }
    }
    return null;
}

function awaitFirstUpdate(ws, timeoutMs) {
    return new Promise((resolve) => {
        let resolved = false;
        const timer = setTimeout(() => {
            if (resolved) return;
            resolved = true;
            resolve(null);
        }, timeoutMs);
        // onUpdate is the NAPI binding's callback-based stream
        ws.onUpdate((err, update) => {
            if (resolved) return;
            resolved = true;
            clearTimeout(timer);
            resolve(err ? null : update);
        }).catch(() => {
            if (resolved) return;
            resolved = true;
            clearTimeout(timer);
            resolve(null);
        });
    });
}

// -----------------------------------------------------------------------
// Matrix
// -----------------------------------------------------------------------

async function runSingleMarket(eid) {
    await caseAsync(`${eid}.subscribe.snapshot_arrives`, async () => {
        const seed = await discoverActiveMarket(eid);
        if (!seed) return 'skip';
        const ws = makeWs(eid);
        await ws.connect();
        try {
            await ws.subscribe(seed.subId);
            const update = await awaitFirstUpdate(ws, 20000);
            if (!update) throw new Error(`${eid}: no WsUpdate within 20s`);
            if (!update.kind) throw new Error(`${eid}: update missing 'kind' discriminator`);
        } finally {
            try { await ws.disconnect(); } catch {}
        }
    });
}

async function runDisconnectClean(eid) {
    await caseAsync(`${eid}.connect_then_disconnect`, async () => {
        const ws = makeWs(eid);
        await ws.connect();
        await ws.disconnect();
        const s = ws.state;
        if (!['Closed', 'Disconnected'].includes(s)) {
            throw new Error(`${eid}: unexpected state after disconnect: ${s}`);
        }
    });
}

async function runBadMarketId(eid) {
    await caseAsync(`${eid}.subscribe.bad_market_id`, async () => {
        const ws = makeWs(eid);
        await ws.connect();
        const bogus = eid === 'kalshi' ? 'OPENPX-NOPE-NOEXIST-9999' : '0';
        try { await ws.subscribe(bogus); } catch { /* either ok */ }
        try { await ws.disconnect(); } catch {}
        // The contract is "no hang, no panic". Reaching here = pass.
    });
}

async function runPolymarketPublicNoAuth() {
    await caseAsync('polymarket.public_no_auth', async () => {
        const seed = await discoverActiveMarket('polymarket');
        if (!seed) return 'skip';
        const ws = new WebSocket('polymarket', {});
        await ws.connect();
        try {
            await ws.subscribe(seed.subId);
            const update = await awaitFirstUpdate(ws, 20000);
            if (!update) throw new Error('polymarket public: no WsUpdate within 20s');
        } finally {
            try { await ws.disconnect(); } catch {}
        }
    });
}

async function runKalshiNoAuthRejected() {
    await caseAsync('kalshi.no_auth_rejected', async () => {
        try {
            new WebSocket('kalshi', {});
        } catch {
            return; // expected
        }
        throw new Error('kalshi WebSocket without credentials must reject');
    });
}

// -----------------------------------------------------------------------
// Driver
// -----------------------------------------------------------------------

async function main() {
    console.log('== TypeScript WS e2e ==');
    for (const eid of ['kalshi', 'polymarket']) {
        console.log(`\n# ${eid}`);
        await runSingleMarket(eid);
        await runDisconnectClean(eid);
        await runBadMarketId(eid);
    }
    console.log('\n# cross-exchange contract');
    await runPolymarketPublicNoAuth();
    await runKalshiNoAuthRejected();

    const summary = {
        pass: results.filter((r) => r.status === 'pass').length,
        fail: results.filter((r) => r.status === 'fail').length,
        skip: results.filter((r) => r.status === 'skip').length,
        total: results.length,
    };
    fs.writeFileSync(RESULTS_PATH, JSON.stringify({ summary, cases: results }, null, 2));
    console.log(`\n${summary.pass} pass / ${summary.fail} fail / ${summary.skip} skip (${summary.total} total)`);
    console.log(`results -> ${path.relative(REPO_ROOT, RESULTS_PATH)}`);
    process.exit(summary.fail === 0 ? 0 : 1);
}

main().catch((e) => {
    console.error('fatal:', e);
    process.exit(2);
});

File: e2e_tests/websockets/python/test_websockets.py
#!/usr/bin/env python3
"""End-to-end WebSocket test for the Python SDK.

Mirrors the Rust matrix in `e2e_tests/websockets/rust/test_websockets.rs` —
exercises the unified `openpx.WebSocket` against both Kalshi and Polymarket
to validate the Python binding hasn't drifted from the unified contract:

  - single-market subscribe → first WsUpdate is a Snapshot
  - take-once `updates()` semantics
  - bad market_id surfaces a session error rather than silent stall
  - public-only Polymarket works without auth
  - Kalshi without auth fails at construction with a clear error

Run from repo root:
    just e2e-python websockets

Output: writes a JSON results document to ../results/python.json and prints
a pass/fail summary, exits non-zero if any case fails.
"""

from __future__ import annotations

import json
import os
import sys
import threading
import time
import traceback
from pathlib import Path
from typing import Any


# Load .env so credentials propagate to the native bindings.
def load_dotenv(p: Path) -> None:
    if not p.exists():
        return
    for line in p.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        k, v = line.split("=", 1)
        os.environ.setdefault(k.strip(), v.strip())


REPO_ROOT = Path(__file__).resolve().parents[3]
load_dotenv(REPO_ROOT / ".env")
key_path = os.environ.get("KALSHI_PRIVATE_KEY_PATH", "")
if key_path and not os.path.isabs(key_path):
    os.environ["KALSHI_PRIVATE_KEY_PATH"] = str(REPO_ROOT / key_path)
RESULTS_DIR = REPO_ROOT / "e2e_tests" / "websockets" / "results"
RESULTS_DIR.mkdir(parents=True, exist_ok=True)
RESULTS_PATH = RESULTS_DIR / "python.json"

from openpx import Exchange, WebSocket  # noqa: E402

KALSHI_CONFIG = {
    "api_key_id": os.environ.get("KALSHI_API_KEY_ID", ""),
    "private_key_path": os.environ.get("KALSHI_PRIVATE_KEY_PATH", ""),
}
POLY_CONFIG = {
    "private_key": os.environ.get("POLYMARKET_PRIVATE_KEY", ""),
    "funder": os.environ.get("POLYMARKET_FUNDER", ""),
    "signature_type": os.environ.get("POLYMARKET_SIGNATURE_TYPE", ""),
    "api_key": os.environ.get("POLYMARKET_API_KEY", ""),
    "api_secret": os.environ.get("POLYMARKET_API_SECRET", ""),
    "api_passphrase": os.environ.get("POLYMARKET_API_PASSPHRASE", ""),
}


def _trim(cfg: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in cfg.items() if v}


def make_exchange(eid: str):
    return Exchange(eid, _trim(KALSHI_CONFIG if eid == "kalshi" else POLY_CONFIG))


def make_ws(eid: str):
    return WebSocket(eid, _trim(KALSHI_CONFIG if eid == "kalshi" else POLY_CONFIG))


# ---------------------------------------------------------------------------
# Result tracking (same shape as e2e_orderbooks/python)
# ---------------------------------------------------------------------------

results: list[dict] = []


def case(name: str, expect: str = "ok"):
    def deco(fn):
        entry = {"name": name, "expect": expect, "status": "pending", "detail": None}
        try:
            fn(entry)
            if entry["status"] == "pending":
                entry["status"] = "fail" if expect == "err" else "pass"
        except AssertionError as e:
            entry["status"] = "fail"
            entry["detail"] = f"assert: {e}\n{traceback.format_exc()}"
        except Exception as e:  # noqa: BLE001
            msg = str(e).lower()
            if expect == "err":
                entry["status"] = "pass"
                entry["detail"] = f"expected err: {e}"
            elif any(t in msg for t in ("rate limit", "429", "timed out", "http error")):
                entry["status"] = "skip"
                entry["detail"] = f"transient: {e}"
            else:
                entry["status"] = "fail"
                entry["detail"] = f"raised: {e}\n{traceback.format_exc()}"
        results.append(entry)
        symbol = {"pass": "✓", "fail": "✗", "skip": "·"}.get(entry["status"], "?")
        print(f"  {symbol} [{entry['status'].upper()}] {name}")
        if entry["status"] == "fail" and entry["detail"]:
            print(f"      → {str(entry['detail'])[:200]}")
        return fn

    return deco


# ---------------------------------------------------------------------------
# Helpers — discover a populated market on each exchange
# ---------------------------------------------------------------------------

def discover_active_market(eid: str):
    """Return (subscribe_id, asset_id) for a live market with a non-empty book.
    Subscribe id == ticker on Kalshi, == token id on Polymarket."""
    ex = make_exchange(eid)
    page = ex.fetch_markets(status="active", limit=50)
    markets = getattr(page, "markets", None) or page[0] if isinstance(page, tuple) else page
    if hasattr(markets, "markets"):
        markets = markets.markets
    for m in list(markets)[:20]:
        outcomes = getattr(m, "outcomes", None) or []
        token = None
        for o in outcomes:
            t = getattr(o, "token_id", None)
            if t:
                token = t
                break
        ticker = getattr(m, "ticker", None)
        asset_id = token or ticker
        if not asset_id:
            continue
        try:
            book = ex.fetch_orderbook(asset_id)
        except Exception:
            continue
        bids = getattr(book, "bids", []) or []
        asks = getattr(book, "asks", []) or []
        if bids or asks:
            sub_id = ticker if eid == "kalshi" else asset_id
            return sub_id, asset_id
    return None


def collect_updates(ws, n: int, timeout_s: float):
    """Collect up to `n` updates within `timeout_s` total. Returns list of
    `kind` strings (Snapshot/Delta/Trade/Fill) for inspection."""
    seen: list[str] = []
    deadline = time.monotonic() + timeout_s

    def reader():
        try:
            for u in ws.updates():
                kind = type(u).__name__
                seen.append(kind)
                if len(seen) >= n or time.monotonic() >= deadline:
                    break
        except Exception:
            pass

    t = threading.Thread(target=reader, daemon=True)
    t.start()
    t.join(timeout=timeout_s)
    return seen


# ---------------------------------------------------------------------------
# Test matrix
# ---------------------------------------------------------------------------

def run_single_market(eid: str):
    @case(f"{eid}.subscribe.snapshot_arrives", "ok")
    def _(entry):
        seed = discover_active_market(eid)
        if seed is None:
            entry["status"] = "skip"
            entry["detail"] = "no live market with populated book"
            return
        sub_id, _asset = seed
        ws = make_ws(eid)
        ws.connect()
        try:
            ws.subscribe(sub_id)
            seen = collect_updates(ws, n=1, timeout_s=20.0)
            assert seen, f"{eid}: no WsUpdate within 20s for {sub_id}"
            assert seen[0] in {"Snapshot", "Delta", "Trade", "Fill"}, f"unknown variant: {seen[0]}"
        finally:
            try:
                ws.disconnect()
            except Exception:
                pass


def run_take_once(eid: str):
    @case(f"{eid}.updates.take_once", "ok")
    def _(entry):
        ws = make_ws(eid)
        first = ws.updates()
        try:
            ws.updates()
        except Exception:
            return  # second call rejecting is the contract
        # If second call returned without raising, the iterator may still be
        # the same single-consumer wrapper; either way the binding must not
        # silently split — accept "raises" or "returns same iterator" but
        # not "returns a new independent receiver".
        first_id = id(first)
        second = ws.updates()
        if id(second) != first_id:
            raise AssertionError(
                f"{eid}: updates() returned independent iterators (split-stream footgun)"
            )


def run_disconnect_clean(eid: str):
    @case(f"{eid}.connect_then_disconnect", "ok")
    def _(entry):
        ws = make_ws(eid)
        ws.connect()
        ws.disconnect()
        # State must be Closed (not Connected/Connecting)
        s = ws.state
        assert s in {"Closed", "Disconnected"}, f"{eid}: state after disconnect: {s}"


def run_bad_market_id(eid: str):
    @case(f"{eid}.subscribe.bad_market_id", "ok")
    def _(entry):
        ws = make_ws(eid)
        ws.connect()
        bogus = "OPENPX-NOPE-NOEXIST-9999" if eid == "kalshi" else "0"
        try:
            ws.subscribe(bogus)
            # Either subscribe raises OR a session event surfaces an error;
            # silent acceptance is also tolerated (some exchanges drop bad
            # ids server-side without emitting an error frame). The contract
            # under test is "no panic, no deadlock", which we already proved.
        except Exception:
            pass
        try:
            ws.disconnect()
        except Exception:
            pass


def run_public_no_auth_polymarket():
    @case("polymarket.public_no_auth", "ok")
    def _(entry):
        seed = discover_active_market("polymarket")
        if seed is None:
            entry["status"] = "skip"
            entry["detail"] = "no live market"
            return
        sub_id, _ = seed
        ws = WebSocket("polymarket", {})  # no creds
        ws.connect()
        try:
            ws.subscribe(sub_id)
            seen = collect_updates(ws, n=1, timeout_s=20.0)
            assert seen, "polymarket public: no WsUpdate within 20s"
        finally:
            try:
                ws.disconnect()
            except Exception:
                pass


def run_no_auth_kalshi_rejected():
    @case("kalshi.no_auth_rejected", "ok")
    def _(entry):
        try:
            WebSocket("kalshi", {})  # constructor must reject
        except Exception:
            return
        raise AssertionError("kalshi WebSocket without credentials must reject")


# ---------------------------------------------------------------------------
# Driver
# ---------------------------------------------------------------------------

def main() -> int:
    print("== Python WS e2e ==")
    for eid in ("kalshi", "polymarket"):
        print(f"\n# {eid}")
        run_single_market(eid)
        run_take_once(eid)
        run_disconnect_clean(eid)
        run_bad_market_id(eid)

    print("\n# cross-exchange contract")
    run_public_no_auth_polymarket()
    run_no_auth_kalshi_rejected()

    summary = {
        "pass": sum(1 for r in results if r["status"] == "pass"),
        "fail": sum(1 for r in results if r["status"] == "fail"),
        "skip": sum(1 for r in results if r["status"] == "skip"),
        "total": len(results),
    }
    RESULTS_PATH.write_text(json.dumps({"summary": summary, "cases": results}, indent=2))
    print(f"\n{summary['pass']} pass / {summary['fail']} fail / {summary['skip']} skip ({summary['total']} total)")
    print(f"results -> {RESULTS_PATH.relative_to(REPO_ROOT)}")
    return 0 if summary["fail"] == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
